java-checks/src/main/java/org/sonar/java/checks/NullShouldNotBeUsedWithOptionalCheck.java

```diff
@@ -0,0 +1,149 @@
+import org.sonar.plugins.java.api.tree.AnnotationTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.ConditionalExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.LambdaExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.ModifiersTree;
+import org.sonar.plugins.java.api.tree.ReturnStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S2789")
+public class NullShouldNotBeUsedWithOptionalCheck extends BaseTreeVisitor implements JavaFileScanner {
+
+  private static final String NULLABLE = "javax.annotation.Nullable";
+
+  private static final String OPTIONAL = "java.util.Optional";
+
+  private JavaFileScannerContext context;
+
+  @Override
+  public void scanFile(final JavaFileScannerContext context) {
+    this.context = context;
+    scan(context.getTree());
+  }
+
+  @Override
+  public void visitMethod(MethodTree method) {
+    if (!method.is(Tree.Kind.CONSTRUCTOR) && returnsOptional(method)) {
+
+      // check that the method is not annotated with @Nullable
+      checkNullableOptionalAnnotation(method.modifiers(), "Methods with an \"Optional\" return type should not be \"@Nullable\".");
+
+      // check that the method does not return "null"
+      method.accept(new ReturnNullVisitor());
+    }
+
+    super.visitMethod(method);
+  }
+
+  @Override
+  public void visitBinaryExpression(BinaryExpressionTree binaryExpression) {
+    // check that an @Optional is not compared to "null"
+    if (binaryExpression.is(Tree.Kind.EQUAL_TO, Tree.Kind.NOT_EQUAL_TO)) {
+      ExpressionTree left = binaryExpression.leftOperand();
+      ExpressionTree right = binaryExpression.rightOperand();
+      if ((isOptional(left) && isNull(right)) || (isNull(left) && isOptional(right))) {
+        context.reportIssue(this, binaryExpression, "Remove this null-check of an \"Optional\".");
+      }
+    }
+
+    super.visitBinaryExpression(binaryExpression);
+  }
+
+  @Override
+  public void visitVariable(VariableTree variable) {
+    if (variable.type().symbolType().is(OPTIONAL)) {
+      checkNullableOptionalAnnotation(variable.modifiers(), "\"Optional\" variables should not be \"@Nullable\".");
+    }
+
+    super.visitVariable(variable);
+  }
+
+  private class ReturnNullVisitor extends BaseTreeVisitor {
+
+    @Override
+    public void visitReturnStatement(ReturnStatementTree returnStatement) {
+      checkNull(returnStatement.expression());
+
+      super.visitReturnStatement(returnStatement);
+    }
+
+    @Override
+    public void visitConditionalExpression(ConditionalExpressionTree conditionalExpression) {
+      checkNull(conditionalExpression.trueExpression());
```

Comment:
This is a bit too strong. You should perform the verification of the true and false expressions only if the type of the expression is Optional. With the current state, the following code will trigger an issue:
```
  public Optional<String> bar(booleab b) {
    Object myObject = b ? null : new Object();
    return Optional.empty();
  }
```

With the following, it should work:
```
 if (conditionalExpression.symbolType().is(OPTIONAL)) {
    checkNull(conditionalExpression.trueExpression());
    checkNull(conditionalExpression.falseExpression());
  }
```